subgraphs/balancer-v2/src/common/metrics.ts

```diff
@@ -0,0 +1,252 @@
+        swapFee.times(supplySidePercentage),
+      );
+      protocol.cumulativeProtocolSideRevenueUSD = protocol.cumulativeProtocolSideRevenueUSD.plus(
+        swapFee.times(protocolSwapPercentage),
+      );
+      protocol.cumulativeTotalRevenueUSD = protocol.cumulativeVolumeUSD.plus(swapFee);
+      financials.dailyProtocolSideRevenueUSD = financials.dailyProtocolSideRevenueUSD.plus(
+        swapFee.times(protocolSwapPercentage),
+      );
+      financials.dailySupplySideRevenueUSD = financials.dailySupplySideRevenueUSD.plus(
+        swapFee.times(supplySidePercentage),
+      );
+      financials.dailyTotalRevenueUSD = financials.dailyTotalRevenueUSD.plus(swapFee);
+      financials.dailyVolumeUSD = financials.dailyVolumeUSD.plus(swapValue);
+    }
+    pool.cumulativeVolumeUSD = pool.cumulativeVolumeUSD.plus(swapValue);
+    protocol.cumulativeVolumeUSD = protocol.cumulativeVolumeUSD.plus(swapValue);
+  }
+  pool.totalValueLockedUSD = newPoolLiquidity;
+  pool.outputTokenPriceUSD = newPoolLiquidity.div(
+    scaleDown(pool.outputTokenSupply, Address.fromString(pool.outputToken)),
+  );
+
+  financials.save();
+  protocol.save();
+  pool.save();
+  updatePoolDailySnapshot(event, pool);
+  updatePoolHourlySnapshot(event, pool);
+}
+
+export function updateTokenPrice(
+  pool: LiquidityPool,
+  tokenA: Address,
+  tokenAAmount: BigInt,
+  tokenAIndex: i32,
+  tokenB: Address,
+  tokenBAmount: BigInt,
+  tokenBIndex: i32,
+  blockNumber: BigInt,
+): void {
+  let hasWeights = !!pool.inputTokenWeights.length;
+
+  let weightTokenB: BigDecimal | null = null;
+  let weightTokenA: BigDecimal | null = null;
+  let tokenAmountIn = scaleDown(tokenAAmount, tokenA);
+  let tokenAmountOut = scaleDown(tokenBAmount, tokenB);
+
+  if (hasWeights) {
+    weightTokenB = pool.inputTokenWeights[tokenBIndex];
+    weightTokenA = pool.inputTokenWeights[tokenAIndex];
+    tokenAmountIn = scaleDown(pool.inputTokenBalances[tokenAIndex], tokenA);
+    tokenAmountOut = scaleDown(pool.inputTokenBalances[tokenBIndex], tokenB);
+  }
+
+  const tokenInfo = calculatePrice(tokenA, tokenAmountIn, weightTokenA, tokenB, tokenAmountOut, weightTokenB);
+
+  if (tokenInfo) {
+    let token = Token.load(tokenInfo.address.toHexString());
+    if (!token) token = new Token(tokenInfo.address.toHexString());
+    const index = tokenInfo.address == tokenB ? tokenBIndex : tokenAIndex;
+    const currentBalance = scaleDown(pool.inputTokenBalances[index], Address.fromString(pool.inputTokens[index]));
+    // We check if current balance multiplied by the price is over 40k USD, if not,
+    // it means that the pool does have too much liquidity, so we fetch the price from
+    // external source
+    if (currentBalance.times(tokenInfo.price).gt(BigDecimal.fromString("40000"))) {
+      token.lastPriceUSD = tokenInfo.price;
+      token.lastPriceBlockNumber = blockNumber;
+      token.save();
+      return;
+    }
+  }
+
+  if (getOrCreateDex().network == "MAINNET") return
```

Comment:
Once the oracle module supports polygon and/or arbitrum, this should be modified